timemachine/cpp/src/potential.cu

```diff
@@ -1,16 +1,112 @@
-#include <iostream>
 #include <memory>
 
 #include "device_buffer.hpp"
 #include "fixed_point.hpp"
 #include "gpu_utils.cuh"
 #include "potential.hpp"
-#include "surreal.cuh"
+
+#include <chrono>
+#include <iostream>
 
 namespace timemachine {
 
 const int Potential::D = 3;
 
+void Potential::execute_batch_host(
+    const int coord_batches,     // Number of batches of coordinates
+    const int N,                 // Number of atoms
+    const int param_batches,     // Number of batches of parameters
+    const int P,                 // Number of parameters
+    const int lambda_batches,    // Number of lambda values
+    const double *h_x,           // [coord_batches, N, 3]
+    const double *h_p,           // [param_batches, P]
+    const double *h_box,         // [coord_batches, 3, 3]
+    const double *lambdas,       // [lambda_batches]
+    unsigned long long *h_du_dx, // [coord_batches, param_batches, lambda_batches, N, 3]
+    unsigned long long *h_du_dp, // [coord_batches, param_batches, lambda_batches, P]
+    unsigned long long *h_du_dl, // [coord_batches, param_batches, lambda_batches, N]
+    unsigned long long *h_u) {   // [coord_batches, param_batches, lambda_batches, N]
+    std::unique_ptr<DeviceBuffer<double>> d_p(nullptr);
+    if (P > 0) {
+        d_p.reset(new DeviceBuffer<double>(param_batches * P));
+        d_p->copy_from(h_p);
+    }
+
+    DeviceBuffer<double> d_box(coord_batches * D * D);
+    d_box.copy_from(h_box);
+
+    DeviceBuffer<double> d_x_buffer(coord_batches * N * D);
+    d_x_buffer.copy_from(h_x);
+
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dx_buffer(nullptr);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dp_buffer(nullptr);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dl_buffer(nullptr);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_u_buffer(nullptr);
+
+    const int K = coord_batches * param_batches * lambda_batches;
+
+    cudaStream_t stream;
+    gpuErrchk(cudaStreamCreate(&stream));
+
+    if (h_du_dx) {
+        d_du_dx_buffer.reset(new DeviceBuffer<unsigned long long>(K * N * D));
+        gpuErrchk(cudaMemsetAsync(d_du_dx_buffer->data, 0, d_du_dx_buffer->size, stream));
+    }
+
+    if (h_du_dp) {
+        d_du_dp_buffer.reset(new DeviceBuffer<unsigned long long>(K * P));
+        gpuErrchk(cudaMemsetAsync(d_du_dp_buffer->data, 0, d_du_dp_buffer->size, stream));
+    }
+
+    if (h_du_dl) {
+        d_du_dl_buffer.reset(new DeviceBuffer<unsigned long long>(K * N));
+        gpuErrchk(cudaMemsetAsync(d_du_dl_buffer->data, 0, d_du_dl_buffer->size, stream));
+    }
+
+    if (h_u) {
+        d_u_buffer.reset(new DeviceBuffer<unsigned long long>(K * N));
+        gpuErrchk(cudaMemsetAsync(d_u_buffer->data, 0, d_u_buffer->size, stream));
+    }
+
+    for (unsigned int i = 0; i < coord_batches; i++) {
+        for (unsigned int j = 0; j < param_batches; j++) {
+            for (unsigned int k = 0; k < lambda_batches; k++) {
```

Comment:
I think this is a good default loop ordering, and if we want to reorder these for any reason from calling code, can reduce this to nested loops over one argument at a time (`coord_batches=1, param_batches=B, lambda_batches=1`) cases.